cmd/outline-ss-server/main.go

```diff
@@ -58,62 +60,49 @@ func init() {
 }
 
-type ssPort struct {
-	tcpListener *net.TCPListener
-	packetConn  net.PacketConn
-	cipherList  service.CipherList
+type Handler interface {
+	NumCiphers() int
+	AddCipher(entry *service.CipherEntry)
+	Handle(ctx context.Context, conn any)
 }
 
-type SSServer struct {
+type connHandler struct {
+	tcpTimeout  time.Duration
 	natTimeout  time.Duration
+	replayCache *service.ReplayCache
 	m           *outlineMetrics
-	replayCache service.ReplayCache
-	ports       map[int]*ssPort
+	ciphers     service.CipherList
 }
 
-func (s *SSServer) startPort(portNum int) error {
-	listener, err := net.ListenTCP("tcp", &net.TCPAddr{Port: portNum})
-	if err != nil {
-		//lint:ignore ST1005 Shadowsocks is capitalized.
-		return fmt.Errorf("Shadowsocks TCP service failed to start on port %v: %w", portNum, err)
-	}
-	logger.Infof("Shadowsocks TCP service listening on %v", listener.Addr().String())
-	packetConn, err := net.ListenUDP("udp", &net.UDPAddr{Port: portNum})
-	if err != nil {
-		//lint:ignore ST1005 Shadowsocks is capitalized.
-		return fmt.Errorf("Shadowsocks UDP service failed to start on port %v: %w", portNum, err)
-	}
-	logger.Infof("Shadowsocks UDP service listening on %v", packetConn.LocalAddr().String())
-	port := &ssPort{tcpListener: listener, packetConn: packetConn, cipherList: service.NewCipherList()}
-	authFunc := service.NewShadowsocksStreamAuthenticator(port.cipherList, &s.replayCache, s.m)
-	// TODO: Register initial data metrics at zero.
-	tcpHandler := service.NewTCPHandler(listener.Addr().String(), authFunc, s.m, tcpReadTimeout)
-	packetHandler := service.NewPacketHandler(s.natTimeout, port.cipherList, s.m)
-	s.ports[portNum] = port
-	go service.StreamServe(service.WrapStreamListener(listener.AcceptTCP), tcpHandler.Handle)
-	go packetHandler.Handle(port.packetConn)
-	return nil
+func (h *connHandler) NumCiphers() int {
+	return h.ciphers.Len()
 }
 
-func (s *SSServer) removePort(portNum int) error {
-	port, ok := s.ports[portNum]
-	if !ok {
-		return fmt.Errorf("port %v doesn't exist", portNum)
-	}
-	tcpErr := port.tcpListener.Close()
-	udpErr := port.packetConn.Close()
-	delete(s.ports, portNum)
-	if tcpErr != nil {
-		//lint:ignore ST1005 Shadowsocks is capitalized.
-		return fmt.Errorf("Shadowsocks TCP service on port %v failed to stop: %w", portNum, tcpErr)
-	}
-	logger.Infof("Shadowsocks TCP service on port %v stopped", portNum)
-	if udpErr != nil {
-		//lint:ignore ST1005 Shadowsocks is capitalized.
-		return fmt.Errorf("Shadowsocks UDP service on port %v failed to stop: %w", portNum, udpErr)
+func (h *connHandler) AddCipher(entry *service.CipherEntry) {
+	h.ciphers.PushBack(entry)
+}
+
+func (h *connHandler) Handle(ctx context.Context, conn any) {
+	switch c := conn.(type) {
+	case transport.StreamConn:
+		authFunc := service.NewShadowsocksStreamAuthenticator(h.ciphers, h.replayCache, h.m)
+		// TODO: Register initial data metrics at zero.
+		tcpHandler := service.NewTCPHandler(c.LocalAddr().String(), authFunc, h.m, h.tcpTimeout)
```

Comment:
We keep rebuilding the same handler. That's the wrong scope. Please create the handler once.

A service should have a list of stream listeners, a stream handler, a list of packet listeners and a packet handler, + the ciphers, which are shared. That properly aligns with the scope.